torch/fx/experimental/graph_gradual_typechecker.py

```diff
@@ -500,3 +522,155 @@ def type_check_node(self, n: Node):
 
         else:
             raise NotImplementedError("Method not yet implemented")
+
+
+@register_refinement_rule(Conv2d)
+@register_refinement_rule(torch.nn.Linear)
+def first_one(n: Node):
+    res = []
+    assert isinstance(n.args[0], Node)
+    arg_type = n.args[0].type
+    if isinstance(arg_type, TensorType) and isinstance(n.type, TensorType):
+        res = [Equality(arg_type.__args__[0], n.type.__args__[0])]
+    return res
+
+# todo needs review for addition. Is this constraint correct?
+@register_refinement_rule(BatchNorm2d)
+@register_refinement_rule(torch.nn.ReLU)
+@register_refinement_rule(torch.nn.AdaptiveAvgPool2d)
+def all_eq(n: Node):
+    res = []
+    assert isinstance(n.args[0], Node)
+    arg_type = n.args[0].type
+    if isinstance(arg_type, TensorType) and isinstance(n.type, TensorType):
+        args1 = arg_type.__args__
+        args2 = n.type.__args__
+        res = [Equality(args1[i], args2[i]) for i in range(len(args1))]
+    return res
+
+@register_refinement_rule(torch.add)
+@register_refinement_rule(operator.add)
+def add_eq(n: Node):
+    res = []
+    if isinstance(n.args[0], Node) and isinstance(n.args[1], Node):
+        arg_type1 = n.args[0].type
+        arg_type2 = n.args[1].type
+        if isinstance(arg_type1, TensorType) and isinstance(arg_type2, TensorType) and isinstance(n.type, TensorType):
+            args1 = arg_type1.__args__
+            args2 = arg_type2.__args__
+            args3 = n.type.__args__
+            r = []
+            for x, y, z in zip(args1, args2, args3):
```

Comment:
since we type-checked the input first, if both operants and n.type are all tensors, then we know that the length must be equal. I think I should eventually put everything in one function instead of calling symbolic inference twice and calling the type-checker once.